### PR TITLE
fix breaking changes for ONNX Runtime Training

### PR DESCRIPTION
**Pull request by ajindal1**
Fixes breaking changes for ONNX Runtime Training.

PR https://github.com/pytorch/pytorch/pull/121102 introduced incompatibility with ORT training because of change in parameter type. Creating a PR to add previous parameter types and verified that it works with ORT training.

Error with current scenario:

```
site-packages/onnxruntime/training/ortmodule/torch_cpp_extensions/cpu/aten_op_executor/aten_op_executor.cc:60:40: error: invalid conversion from ‘const DLManagedTensor*’ to ‘DLManagedTensor*’ [-fpermissive]
at::Tensor tensor = at::fromDLPack(dlpack);

site-packages/torch/include/ATen/DLConvertor.h:15:46: note:   initializing argument 1 of ‘at::Tensor at::fromDLPack(DLManagedTensor*)’
TORCH_API Tensor fromDLPack(DLManagedTensor* src);
```

### CODE REVIEW

**Comment by ajindal1** [comment]
@cyyever @Skylion007 any thoughts on this?

**Comment by cyyever** [comment]
@ajindal1 LGTM, hope that tests pass.

File: aten/src/ATen/DLConvertor.cpp [hunk 1]
```diff
@@ -298,6 +298,16 @@ Tensor fromDLPack(DLManagedTensor* src) {
...
 }
 
+Tensor fromDLPack(const DLManagedTensor* src) {
+  auto deleter = [src](void* self) {
+    if (src->deleter) {
+      // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
+      src->deleter(const_cast<DLManagedTensor*>(src));
+    }
+  };
+  return fromDLPack(src, std::move(deleter));
```
Comment (malfet) [inline aten/src/ATen/DLConvertor.cpp:308]: Do not copy the code, just cost cast a top level wrapper
```suggestion
      // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
     return from DLPack(const_cast<DLManagedTensor*>(src));
```

File: aten/src/ATen/DLConvertor.h [hunk 2]
```diff
@@ -13,8 +13,11 @@ namespace at {
 TORCH_API ScalarType toScalarType(const DLDataType& dtype);
 TORCH_API DLManagedTensor* toDLPack(const Tensor& src);
 TORCH_API Tensor fromDLPack(DLManagedTensor* src);
+TORCH_API Tensor fromDLPack(const DLManagedTensor* src);
```
Comment (malfet) [inline aten/src/ATen/DLConvertor.h:16]: As those functions are semantically incorrect, let's mark them as deprecated an implement as headers to casts to a const variants
```suggestion
C10_DEPRECATED_MESSAGE("Please migrate to a non-const variant")
inline Tensor fromDLPack(const DLManagedTensor* src) { return fromDLPack(const_cast< DLManagedTensor*>(src); }
```

**Review by malfet — CHANGES_REQUESTED** [review]
Why do the need a duplicate implementation rather than do a `const_cast` in the header
(But also mark them deprecated so that there will be no use of such functions in Torch

File: aten/src/ATen/DLConvertor.h (hunk 2 above)
Comment (ajindal1) [reply aten/src/ATen/DLConvertor.h]: Thanks, I have modified just the header file as you suggested, which is minimal changes and makes sense to me. I have started a run on my end to verify it works. Will confirm it by EOD.

**Comment by cyyever** [comment]
@ajindal1 The parameter is const_casted in the function body so that it shouldn't be taken as const. Is it possible to patch aten_op_executor.cc to use the non-const version?

**Comment by ajindal1** [comment]
@cyyever I don't completely understand your comment, can you please add more information. I am also unable to locate aten_op_executor.cc file you mentioned.

**Comment by wschin** [comment]
> @cyyever I don't completely understand your comment, can you please add more information. I am also unable to locate aten_op_executor.cc file you mentioned.

I guess this mean to call `const_cast` inside aten_op_executor.cc to match whatever the new behavior in PyTorch C++. Not sure if it's doable. On the other hand, I'd prefer PyTorch to maintain BC whenever possible.

**Comment by ajindal1** [comment]
@cyyever we have also added the const cast change in onnxruntime repository. However, for backward compatibility we will need this change here as well.

**Comment by cyyever** [comment]
@ajindal1 Thank you!

**Comment by ajindal1** [comment]
@malfet does everything look good to you now?

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
@pytorchbbot merge -if "I've asked to ignore that failure, don't I?"

**Comment by ajindal1** [comment]
Validated with 2.3.

**Pull request closed without merging**